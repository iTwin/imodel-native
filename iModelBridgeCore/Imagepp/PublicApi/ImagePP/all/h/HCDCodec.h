//:>--------------------------------------------------------------------------------------+
//:>
//:>     $Source: PublicApi/ImagePP/all/h/HCDCodec.h $
//:>
//:>  $Copyright: (c) 2015 Bentley Systems, Incorporated. All rights reserved. $
//:>
//:>+--------------------------------------------------------------------------------------
//-----------------------------------------------------------------------------
// Class : HCDCodec
//-----------------------------------------------------------------------------
// General class for codecs.
//-----------------------------------------------------------------------------

#pragma once

#include <Imagepp/all/h/HFCExclusiveKey.h>

BEGIN_IMAGEPP_NAMESPACE

class HCDPacketRLE;

// Binary optimization: Speed up read/write when codec support direct decompression/compression to/from RLE format.
class HNOVTABLEINIT HCDCodecRLEInterface
    {
public:
    virtual void     DecompressSubsetToRLE(const void* pi_pInData, size_t pi_InDataSize, HFCPtr<HCDPacketRLE>& pio_rpRLEPacket) = 0;
    virtual size_t   CompressSubsetFromRLE(HFCPtr<HCDPacketRLE> const& pi_rpPacketRLE, void* po_pOutBuffer, size_t po_OutBufferSize) = 0;
    };

class HNOVTABLEINIT HCDCodec : public HFCShareableObject<HCDCodec>,
    public HFCExclusiveKey     // to have access to claim/release instead of including m_Key and adding the methods
    {
    HDECLARE_BASECLASS_ID(HCDCodecId_Base)

public:

    // codec states enumeration
    enum State
        {
        STATE_NONE,
        STATE_COMPRESS,
        STATE_DECOMPRESS
        };

    // destructor
    virtual         ~HCDCodec();

    // compression
    virtual size_t          CompressSubset(const void* pi_pInData,
                                           size_t pi_InDataSize,
                                           void* po_pOutBuffer,
                                           size_t pi_OutBufferSize) = 0;

    virtual size_t          DecompressSubset(const void* pi_pInData,
                                             size_t pi_InDataSize,
                                             void* po_pOutBuffer,
                                             size_t pi_OutBufferSize) = 0;

    State                   GetCurrentState() const;
    virtual void            Reset();

    virtual HCDCodec*       Clone() const = 0;

    virtual size_t          GetDataSize() const = 0;

    // subset
    virtual size_t          GetMinimumSubsetSize() const = 0;
    virtual size_t          GetSubsetMaxCompressedSize() const = 0;
    virtual void            SetSubsetSize(size_t pi_Size) = 0;

    // others
    virtual double          GetEstimatedCompressionRatio() const;
    virtual bool            HasRandomAccess() const;
    virtual bool            IsCodecImage() const;
    virtual bool            IsCodecVector() const;
    virtual const WString&  GetName() const;

    virtual HCDCodecRLEInterface* 
                            GetRLEInterface();

protected:

    // constructors
    HCDCodec(const WString& pi_rCodecName);
    HCDCodec(const HCDCodec& pi_rObj);


    // compression
    void                    SetCurrentState(State pi_State);

    WString         m_CodecName;

private:

    State          m_CurrentState;
    };

END_IMAGEPP_NAMESPACE