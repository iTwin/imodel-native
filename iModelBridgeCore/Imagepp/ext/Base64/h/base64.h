#ifndef _base64_h_
#define _base64_h_

#include <iostream>

namespace b64
{
        bool encode(std::istream& input, std::ostream& output);
        bool decode(std::istream& input, std::ostream& output);
};

#endif
