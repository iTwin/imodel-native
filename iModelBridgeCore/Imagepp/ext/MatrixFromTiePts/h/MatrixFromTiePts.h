/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See COPYRIGHT.md in the repository root for full copyright notice.
*--------------------------------------------------------------------------------------------*/

#pragma once


#define TRANSFO_NONE     -1 /* No transformation model */
#define TRANSFO_TRAN_2D   0 /* Translation en x et y */
#define TRANSFO_SIMI_2D   1 /* Translation x,y + Rotation + Echelle */
#define TRANSFO_POLY_2D_1 2 /* Polynomiale deg 1 */
#define TRANSFO_AFFI_2D   3 /* Tran. x,y + Rotation + Echelle + Axes */
#define TRANSFO_PROJ_2D   4 /* Projection d'un plan sur un autre */
#define TRANSFO_POLY_2D_2 5 /* Polynomiale deg 2 */
#define TRANSFO_POLY_2D_3 6 /* Polynomiale deg 3 */
#define TRANSFO_POLY_2D_4 7 /* Polynomiale deg 4 */ /* A tester ??? */
#define TRANSFO_POLY_2D_5 8 /* Polynomiale deg 5 */ /* A tester ??? */
#define TRANSFO_MAT_2D    9 /* Matrice de transformation */
#define TRANSFO_HELM_2D   10
#define TRANSFO_TPS_2D    11
#define TRANSFO_MAX       12

#define TRF_ERROR_MAX_ITERATIONS    -1
#define TRF_ERROR_NO_MEMORY         -2
#define TRF_ERROR_BAD_NUMBER_POINTS -3
#define TRF_ERROR_MATRIX            -4


long GetTransfoMatrixFromScaleAndTiePts (double          po_pMatrix[4][4],
                                         unsigned short  pi_NbVal_GeoTiePoint,
                                         double*         pi_pVal_GeoTiePoint,
                                         unsigned short  pi_NbVal_GeoPixelScale,
                                         double*         pi_pVal_GeoPixelScale);

long GetProjectiveTransfoMatrixFromScaleAndTiePts(double         po_pMatrix[4][4],
                                                  unsigned short pi_NbVal_GeoTiePoint,
                                                  double*        pi_pVal_GeoTiePoint);

long GetAffineTransfoMatrixFromScaleAndTiePts(double         po_pMatrix[4][4],
                                              unsigned short pi_NbVal_GeoTiePoint,
                                              double*        pi_pVal_GeoTiePoint);

long GetTransfoMatrixFromTiePts(double           po_pMatrix[4][4],
                                unsigned short   pi_NbVal_GeoTiePoint,
                                const double*    pi_pVal_GeoTiePoint,
                                long             pi_ModelType);

