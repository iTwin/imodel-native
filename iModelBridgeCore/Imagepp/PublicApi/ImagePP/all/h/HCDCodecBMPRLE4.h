//:>--------------------------------------------------------------------------------------+
//:>
//:>
//:>  Copyright (c) Bentley Systems, Incorporated. All rights reserved.
//:>
//:>+--------------------------------------------------------------------------------------
// Class : HCDCodecBMPRLE8
//-----------------------------------------------------------------------------
// RLE8 codec lass.
//-----------------------------------------------------------------------------
#pragma once

#include "HCDCodecImage.h"

BEGIN_IMAGEPP_NAMESPACE

class HCDCodecBMPRLE4 : public HCDCodecImage
    {
    HDECLARE_CLASS_ID(HCDCodecId_BMPRLE4, HCDCodecImage)

public:

    IMAGEPP_EXPORT                 HCDCodecBMPRLE4();

    HCDCodecBMPRLE4(size_t pi_Width,
                    size_t pi_Height);

    HCDCodecBMPRLE4(const HCDCodecBMPRLE4& pi_rObj);

    ~HCDCodecBMPRLE4();

    size_t          CompressSubset(const void* pi_pInData,
                                   size_t     pi_InDataSize,
                                   void*      po_pOutBuffer,
                                   size_t     pi_OutBufferSize) override;

    size_t          DecompressSubset(const void* pi_pInData,
                                     size_t       pi_InDataSize,
                                     void*        po_pOutBuffer,
                                     size_t       pi_OutBufferSize) override;

    bool           HasLineAccess() const override;

    virtual HCDCodec* Clone() const override;

    bool           IsBitsPerPixelSupported(size_t pi_Bits) const override;

    size_t          GetSubsetMaxCompressedSize() const override;

protected:

private:
    };

END_IMAGEPP_NAMESPACE
