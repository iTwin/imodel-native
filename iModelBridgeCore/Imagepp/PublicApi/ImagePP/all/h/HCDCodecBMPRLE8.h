//:>--------------------------------------------------------------------------------------+
//:>
//:>     $Source: PublicApi/ImagePP/all/h/HCDCodecBMPRLE8.h $
//:>
//:>  $Copyright: (c) 2015 Bentley Systems, Incorporated. All rights reserved. $
//:>
//:>+--------------------------------------------------------------------------------------
// Class : HCDCodecBMPRLE8
//-----------------------------------------------------------------------------
// RLE8 codec lass.
//-----------------------------------------------------------------------------
#pragma once

#include "HCDCodecImage.h"

BEGIN_IMAGEPP_NAMESPACE

class HCDCodecBMPRLE8 : public HCDCodecImage
    {
    HDECLARE_CLASS_ID(HCDCodecId_BMPRLE8, HCDCodecImage)

public:

    IMAGEPP_EXPORT                 HCDCodecBMPRLE8();

    HCDCodecBMPRLE8(  size_t pi_Width,
                      size_t pi_Height);

    HCDCodecBMPRLE8(const HCDCodecBMPRLE8& pi_rObj);

    ~HCDCodecBMPRLE8();

    size_t          CompressSubset(const void* pi_pInData,
                                   size_t pi_InDataSize,
                                   void* po_pOutBuffer,
                                   size_t pi_OutBufferSize);

    size_t          DecompressSubset(const void* pi_pInData,
                                     size_t pi_InDataSize,
                                     void* po_pOutBuffer,
                                     size_t pi_OutBufferSize);

    bool           HasLineAccess() const;

    virtual HCDCodec* Clone() const override;

    bool           IsBitsPerPixelSupported(size_t pi_Bits) const;

    size_t          GetSubsetMaxCompressedSize() const;

protected:

private:
    };

END_IMAGEPP_NAMESPACE