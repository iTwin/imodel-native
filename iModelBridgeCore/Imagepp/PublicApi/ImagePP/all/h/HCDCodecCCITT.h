//:>--------------------------------------------------------------------------------------+
//:>
//:>
//:>  Copyright (c) Bentley Systems, Incorporated. All rights reserved.
//:>
//:>+--------------------------------------------------------------------------------------
//-----------------------------------------------------------------------------
// Class : HCDCodecCCITT
//-----------------------------------------------------------------------------
// CCITT codec lass.
//-----------------------------------------------------------------------------

#pragma once

#include "HCDCodecImage.h"

BEGIN_IMAGEPP_NAMESPACE

class HCDCodecCCITT : public HCDCodecImage
    {
    HDECLARE_CLASS_ID(HCDCodecId_CCITT, HCDCodecImage)

public:

    ~HCDCodecCCITT();

    bool           IsBitsPerPixelSupported(size_t pi_Bits) const override;

    HCDCodecCCITT(const HCDCodecCCITT& pi_rObj);

    size_t          GetSubsetMaxCompressedSize() const override;

    double GetEstimatedCompressionRatio() const override;

    IMAGEPP_EXPORT void     SetPhotometric(uint16_t pi_Photo);
    uint16_t GetPhotometric() const;

    IMAGEPP_EXPORT void     SetBitRevTable(bool pi_Reverse);

protected:
    
    virtual HCDCodec* Clone() const override = 0;

    HCDCodecCCITT();

    HCDCodecCCITT(size_t pi_Width,
                  size_t pi_Height);

    bool           m_bitrevtable;
    uint16_t m_photometric;
    };

END_IMAGEPP_NAMESPACE
