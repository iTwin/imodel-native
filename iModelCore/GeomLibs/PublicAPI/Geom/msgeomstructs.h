/*--------------------------------------------------------------------------------------+
|
|     $Source: PublicAPI/Geom/msgeomstructs.h $
|
|  $Copyright: (c) 2011 Bentley Systems, Incorporated. All rights reserved. $
|
+--------------------------------------------------------------------------------------*/
#pragma once

/*__BENTLEY_INTERNAL_ONLY__*/

#include <Geom/GeomApi.h>


