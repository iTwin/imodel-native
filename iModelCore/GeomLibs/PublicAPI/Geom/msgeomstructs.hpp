/*--------------------------------------------------------------------------------------+
|
|     $Source: PublicAPI/Geom/msgeomstructs.hpp $
|
|  $Copyright: (c) 2009 Bentley Systems, Incorporated. All rights reserved. $
|
+--------------------------------------------------------------------------------------*/
#pragma once
/*__BENTLEY_INTERNAL_ONLY__*/
#include <Geom/GeomApi.h>
