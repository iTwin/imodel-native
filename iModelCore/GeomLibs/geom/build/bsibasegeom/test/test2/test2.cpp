/*--------------------------------------------------------------------------------------+
|
|  Copyright (c) Bentley Systems, Incorporated. All rights reserved.
|
+--------------------------------------------------------------------------------------*/
// test1.cpp : Defines the entry point for the console application.
//

#include <stdio.h>
#include <msgeomstructs.hpp>
#include <msgeomstructs.h>
#include <printfuncs.h>
static double s_pi = 4.0 * atan (1.0);

// Standard view data as of July 12, 2005.
static RotMatrix s_standardViewMatrix[8] = {
    {   // Standard view matrix 1 (zero based 0)
         1.000000000000000,  0.00000000000000000,  0.00000000000000000,
         0.000000000000000,  1.00000000000000000,  0.00000000000000000,
         0.000000000000000,  0.00000000000000000,  1.00000000000000000,
    },
    {   // Standard view matrix 2 (zero based 1)
         1.000000000000000,  0.00000000000000000,  0.00000000000000000,
         0.000000000000000, -1.00000000000000000,  0.00000000000000000,
         0.000000000000000,  0.00000000000000000, -1.00000000000000000,
    },
    {   // Standard view matrix 3 (zero based 2)
         0.000000000000000, -1.00000000000000000,  0.00000000000000000,
         0.000000000000000,  0.00000000000000000,  1.00000000000000000,
        -1.000000000000000,  0.00000000000000000,  0.00000000000000000,
    },
    {   // Standard view matrix 4 (zero based 3)
         0.000000000000000,  1.00000000000000000,  0.00000000000000000,
         0.000000000000000,  0.00000000000000000,  1.00000000000000000,
         1.000000000000000,  0.00000000000000000,  0.00000000000000000,
    },
    {   // Standard view matrix 5 (zero based 4)
         1.000000000000000,  0.00000000000000000,  0.00000000000000000,
         0.000000000000000,  0.00000000000000000,  1.00000000000000000,
         0.000000000000000, -1.00000000000000000,  0.00000000000000000,
    },
    {   // Standard view matrix 6 (zero based 5)
        -1.000000000000000,  0.00000000000000000,  0.00000000000000000,
         0.000000000000000,  0.00000000000000000,  1.00000000000000000,
         0.000000000000000,  1.00000000000000000,  0.00000000000000000,
    },
    {   // Standard view matrix 7 (zero based 6)
         0.707106781186548, -0.70710678118654757,  0.00000000000000000,
         0.408248290463863,  0.40824829046386302,  0.81649658092772603,
        -0.577350269189626, -0.57735026918962573,  0.57735026918962573,
    },
    {   // Standard view matrix 8 (zero based 7)
         0.707106781186548,  0.70710678118654757,  0.00000000000000000,
        -0.408248290463863,  0.40824829046386302,  0.81649658092772603,
         0.577350269189626, -0.57735026918962573,  0.57735026918962573,
    },
    };
void printStandardViews ()
    {
    RotMatrix matrix;
    for (int viewId = 1; viewId <= 8; viewId++)
        {
        bsiDMatrix3d_getStandardRotation ((DMatrix3d*)&matrix, viewId);
        printf ("    {   // Standard view matrix %d (zero based %d)\n", viewId, viewId - 1);
        for (int i = 0; i < 3; i++)
            {
            printf ("      %20.15lf, %20.17lf, %20.17lf,\n",
                    matrix.form3d[i][0],
                    matrix.form3d[i][1],
                    matrix.form3d[i][2]);
            }
        printf ("    },\n");
        }
    }

void compareStandardViews ()
    {
    RotMatrix matrix, localMatrix;
    for (int viewId = 1; viewId <= 8; viewId++)
        {
        bsiDMatrix3d_getStandardRotation ((DMatrix3d*)&matrix, viewId);
        localMatrix = s_standardViewMatrix[viewId-1];
        checkRotMatrix (&matrix, &localMatrix, "bsiDMatrix3d_getStandardRotation vs static");
        }
    }

void checkMat (RotMatrix matrix, double thetaXA, double thetaYA, double thetaZA, double sA)
    {
    double thetaX, thetaY, thetaZ, s;
    checkTrue (bsiRotMatrix_isXRotationYRotationZRotationScale( &matrix, &thetaX, &thetaY, &thetaZ, &s),
                "isXYZs");
    checkDouble (thetaX, thetaXA, "isXYZs:thetaX");
    checkDouble (thetaY, thetaYA, "isXYZs:thetaY");
    checkDouble (thetaZ, thetaZA, "isXYZs:thetaZ");
    checkDouble (s, sA, "isXYZs:s");
    }

void checkMatrixDecomp ()
    {
    RotMatrix matrix;
    bsiRotMatrix_initFromRowValues (&matrix, 1,0,0, 0,1,0, 0,0,1);
    checkMat (matrix, 0.0, 0.0, 0.0, 1.0);
    bsiRotMatrix_initFromRowValues (&matrix, -1,0,0, 0,-1,0, 0,0,-1);
    checkMat (matrix, 0.0, 0.0, 0.0, -1.0);

    double alpha0 = -0.25;
    double alpha1 =  0.25;
    double dAlpha =  0.25;

    for (double alphaX = alpha0; alphaX <= alpha1; alphaX += dAlpha)
        {
        for (double alphaY = alpha0; alphaY <= alpha1; alphaY += dAlpha)
            {
            for (double alphaZ = alpha0; alphaZ <= alpha1; alphaZ += dAlpha)
                {
                for (double s = -3.0; s <= 3.0; s += 2.0)
                    {
                    bsiRotMatrix_rotateByPrincipleAngles (&matrix, NULL, alphaX, alphaY, alphaZ);
                    bsiRotMatrix_scaleColumns (&matrix, &matrix, s, s, s);
                    checkMat (matrix, alphaX, alphaY, alphaZ, s);
                    }
                }
            }
        }
    }

int main(int argc, char * argv[])
    {
    initErrorTracking (NULL, argc, argv);

    //printStandardViews ();
    compareStandardViews ();
    checkMatrixDecomp ();
    return getExitStatus();
    }
