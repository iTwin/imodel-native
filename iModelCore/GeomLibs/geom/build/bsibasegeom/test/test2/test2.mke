#--------------------------------------------------------------------------------------
#
#  Copyright (c) Bentley Systems, Incorporated. All rights reserved.
#
#--------------------------------------------------------------------------------------
TESTAPP_NAME=test2

%include    $(_MakeFilePath)/../testapp.mki
