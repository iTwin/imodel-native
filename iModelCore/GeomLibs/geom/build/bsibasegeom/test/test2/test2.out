 geomlibs test environment -- command line arguments:
 Tail of argv[0] = bsigeomtest2.exe
<OK name="bsiDMatrix3d_getStandardRotation vs static"<row0>1,0,0</row0><row1>0,1,0</row1><row2>0,0,1]</row2>/>
<OK name="bsiDMatrix3d_getStandardRotation vs static"<row0>1,0,0</row0><row1>0,-1,0</row1><row2>0,0,-1]</row2>/>
<OK name="bsiDMatrix3d_getStandardRotation vs static"<row0>0,-1,0</row0><row1>0,0,1</row1><row2>-1,0,0]</row2>/>
<OK name="bsiDMatrix3d_getStandardRotation vs static"<row0>0,1,0</row0><row1>0,0,1</row1><row2>1,0,0]</row2>/>
<OK name="bsiDMatrix3d_getStandardRotation vs static"<row0>1,0,0</row0><row1>0,0,1</row1><row2>0,-1,0]</row2>/>
<OK name="bsiDMatrix3d_getStandardRotation vs static"<row0>-1,0,0</row0><row1>0,0,1</row1><row2>0,1,0]</row2>/>
<OK name="bsiDMatrix3d_getStandardRotation vs static"<row0>0.707107,-0.707107,0</row0><row1>0.408248,0.408248,0.816497</row1><row2>-0.57735,-0.57735,0.57735]</row2>/>
<OK name="bsiDMatrix3d_getStandardRotation vs static"<row0>0.707107,0.707107,0</row0><row1>-0.408248,0.408248,0.816497</row1><row2>0.57735,-0.57735,0.57735]</row2>/>
Exit with 0 errors, 8 OK
